### ISSUE TITLE
Have you considered implementing ABY2.0 protocol?

### ISSUE THREAD

**Issue by BeStrongok** [secretflow/spu#125]
Hi, :) 
I'm noticed that the [**ABY2.0**](https://www.usenix.org/system/files/sec21-patra.pdf) protocol has constant round in online communication, which is very effective for Secure Two-Party Computation.
Have you considered implementing this protocol? When i'm benchmarking **Cheetah**, i don't think it's fast enougth, maybe this protocol is more efficiently?
BTW, because SPU has JIT compilation, i'm curious about how to divide offline & online process.

**Comment by fionser** [comment]
A quick response is no.

- For the current point, I am seeing the advantage of using ABY2.0. Especially for large arithmetic computation (matmul, conv2d etc). 
- Also, GC demands a significantly larger communication overhead. I am not sure whether the SPU team will have any idea to implement GC.

**Comment by icavan** [comment]
> Hi, :) I'm noticed that the [**ABY2.0**](https://www.usenix.org/system/files/sec21-patra.pdf) protocol has constant round in online communication, which is very effective for Secure Two-Party Computation. Have you considered implementing this protocol? When i'm benchmarking **Cheetah**, i don't think it's fast enougth, maybe this protocol is more efficiently? BTW, because SPU has JIT compilation, i'm curious about how to divide offline & online process.

hi @BeStrongok  , as @fionser  point out, I do not think `ABY 2.0` deserve the implementation since it will be slower than `Cheetah`. 2PC mpc is still quite challenging, if you've tested the `cheetah LR` performance, you'll find it at least 10x faster than the FATE `3pc` Hetero LR, which is already quite amazing since Cheetah does not require a trusted `Arbitor` role.

hi @fionser , GC requires too much communication overhead, we do have an internal version but not yet been widely used in real business. We might consider adding one for research purposes.

**Comment by icavan** [comment]
> Hi, :) I'm noticed that the [**ABY2.0**](https://www.usenix.org/system/files/sec21-patra.pdf) protocol has constant round in online communication, which is very effective for Secure Two-Party Computation. Have you considered implementing this protocol? When i'm benchmarking **Cheetah**, i don't think it's fast enougth, maybe this protocol is more efficiently? BTW, because SPU has JIT compilation, i'm curious about how to divide offline & online process.

BTW, the challenging part of 2PC protocol is the offline stage. Hence, `constant round` in the online stage does not really make a big difference.

Currently, the offline & online stages are merged together. We plan to add a Just-In-Time offline correlated randomness generator that tries to pre-compute the offline stuff beforehand. We believe it will help improve the performance of `2PC applications`. Stay tuned.

**Comment by BeStrongok** [comment]
> > Hi, :) I'm noticed that the [**ABY2.0**](https://www.usenix.org/system/files/sec21-patra.pdf) protocol has constant round in online communication, which is very effective for Secure Two-Party Computation. Have you considered implementing this protocol? When i'm benchmarking **Cheetah**, i don't think it's fast enougth, maybe this protocol is more efficiently? BTW, because SPU has JIT compilation, i'm curious about how to divide offline & online process.
> 
> BTW, the challenging part of 2PC protocol is the offline stage. Hence, `constant round` in the online stage does not really make a big difference.
> 
> Currently, the offline & online stages are merged together. We plan to add a Just-In-Time offline correlated randomness generator that tries to pre-compute the offline stuff beforehand. We believe it will help improve the performance of `2PC applications`. Stay tuned.

Thanks for quickly reply.  @fionser @icavan :)
I think constant round in the online stage is really helpful in secure network inference, the pre-compute offline triples can be saved in database and be used in-time, especially when the network latency is very high.
I will continue to learn ** Cheetah** protocol, many principles and details are not yet fully understood.

**Comment by icavan** [comment]
> > > Hi, :) I'm noticed that the [**ABY2.0**](https://www.usenix.org/system/files/sec21-patra.pdf) protocol has constant round in online communication, which is very effective for Secure Two-Party Computation. Have you considered implementing this protocol? When i'm benchmarking **Cheetah**, i don't think it's fast enougth, maybe this protocol is more efficiently? BTW, because SPU has JIT compilation, i'm curious about how to divide offline & online process.
> > 
> > 
> > BTW, the challenging part of 2PC protocol is the offline stage. Hence, `constant round` in the online stage does not really make a big difference.
> > Currently, the offline & online stages are merged together. We plan to add a Just-In-Time offline correlated randomness generator that tries to pre-compute the offline stuff beforehand. We believe it will help improve the performance of `2PC applications`. Stay tuned.
> 
> Thanks for quickly reply. @fionser @icavan :) I think constant round in the online stage is really helpful in secure network inference, the pre-compute offline triples can be saved in database and be used in-time, especially when the network latency is very high. I will continue to learn **Cheetah** protocol, many principles and details are not yet fully understood.

Yes, constant round solutions like FHE/GC-style are appreciated, especially in inference scenarios. Users need to check what's the real bottleneck: computation, bandwidth, or latency. Hope you get your answer in your cases.

**Comment by BeStrongok** [comment]
> implementation

Thanks for your guidance!!  More research i need to do, my case is high latency, so constant round is really an important factor. :)

**Comment by BeStrongok** [comment]
> > > > Hi, :) I'm noticed that the [**ABY2.0**](https://www.usenix.org/system/files/sec21-patra.pdf) protocol has constant round in online communication, which is very effective for Secure Two-Party Computation. Have you considered implementing this protocol? When i'm benchmarking **Cheetah**, i don't think it's fast enougth, maybe this protocol is more efficiently? BTW, because SPU has JIT compilation, i'm curious about how to divide offline & online process.
> > > 
> > > 
> > > BTW, the challenging part of 2PC protocol is the offline stage. Hence, `constant round` in the online stage does not really make a big difference.
> > > Currently, the offline & online stages are merged together. We plan to add a Just-In-Time offline correlated randomness generator that tries to pre-compute the offline stuff beforehand. We believe it will help improve the performance of `2PC applications`. Stay tuned.
> > 
> > 
> > Thanks for quickly reply. @fionser @icavan :) I think constant round in the online stage is really helpful in secure network inference, the pre-compute offline triples can be saved in database and be used in-time, especially when the network latency is very high. I will continue to learn **Cheetah** protocol, many principles and details are not yet fully understood.
> 
> Yes, constant round solutions like FHE/GC-style are appreciated, especially in inference scenarios. Users need to check what's the real bottleneck: computation, bandwidth, or latency. Hope you get your answer in your cases.

Thanks for your guidance, more research i need to do. My case is high latency, so constant round is really what i need. :)

**Comment by BeStrongok** [comment]
> A quick response is no.
> 
> * For the current point, I am not seeing the advantage of using ABY2.0. Especially for large arithmetic computation (matmul, conv2d etc).
> * Also, GC demands a significantly larger communication overhead. I am not sure whether the SPU team will have any idea to implement GC.

Thank you for point out this factors, GC is really expensive in large arithmetic computation, such as network communication, what confuses me is the high latency~~

**Comment by anakinxc** [comment]
Hi @BeStrongok 

One question here, you mentioned you were benchmarking Cheetah, may I ask with which version of SPU?

Thanks

**Comment by BeStrongok** [comment]
> Hi @BeStrongok
> 
> One question here, you mentioned you were benchmarking Cheetah, may I ask with which version of SPU?
> 
> Thanks

I have opened an issue about the performance of Cheetah [here](https://github.com/secretflow/secretflow/issues/148), maybe the SPU Team made some improvement in the subsequent version.

**Comment by anakinxc** [comment]
> > Hi @BeStrongok
> > One question here, you mentioned you were benchmarking Cheetah, may I ask with which version of SPU?
> > Thanks
> 
> I have opened an issue about the performance of Cheetah [here](https://github.com/secretflow/secretflow/issues/148), maybe the SPU Team made some improvement in the subsequent version.

Hi @BeStrongok,

That is pretty old.

We have made lots of improvements to both Cheetah(credits to @fionser) and other thingy. I would recommend you give it another try :P

**Comment by BeStrongok** [comment]
> > > Hi @BeStrongok
> > > One question here, you mentioned you were benchmarking Cheetah, may I ask with which version of SPU?
> > > Thanks
> > 
> > 
> > I have opened an issue about the performance of Cheetah [here](https://github.com/secretflow/secretflow/issues/148), maybe the SPU Team made some improvement in the subsequent version.
> 
> Hi @BeStrongok,
> 
> That is pretty old.
> 
> We have made lots of improvements to both Cheetah(credits to @fionser) and other thingy. I would recommend you give it another try :P

Thanks, it's pretty cool! I will try new version!

**Comment by BeStrongok** [comment]
Hi, SPU Team: 
I'm reading the paper of [ABY2.0](https://www.usenix.org/system/files/sec21-patra.pdf), and find that this paper inroduced some improved primitives, such as sec. 5.3 Depth-Optimized Circuits and sec. 5.5 Truncation.
As for Depth-Optimized Circuits, the depth of circuits can be reduced from 6 to 3 in 64 bit length through 3- and 4-input AND/MUL gates.
As for Truncation, it refers [SecureML](https://ieeexplore.ieee.org/stamp/stamp.jsp?tp=&arnumber=7958569) and do not influence the communication round in multiplication.
I'm not sure these primitives have been implemented in SPU, if so, can you help me point out the corresponding code? I want to learn it, thanks very much!

**Comment by anakinxc** [comment]
Hi @BeStrongok 

SPU has implemented PPA [here](https://github.com/secretflow/spu/blob/ad9d3b34ea9e0665c7fd5328ca031c75aba6a161/libspu/mpc/aby3/conversion.cc#L127).

For truncation, we have lots of [truncation kernels](https://github.com/search?q=repo%3Asecretflow%2Fspu%20TruncAKernel&type=code). SecureML's you can find it [here](https://github.com/secretflow/spu/blob/ad9d3b34ea9e0665c7fd5328ca031c75aba6a161/libspu/mpc/semi2k/arithmetic.cc#L251)

Thanks

**Comment by BeStrongok** [comment]
> 

Thanks for pointing these codes, it's really helpful! But it seems not the 3- and 4-input AND/MUL gates introduced by `ABY2`. 
Maybe i'm wrong, i will learn your code carefully.

**Comment by rivertalk** [comment]
> Thanks for pointing these codes, it's really helpful! But it seems not the 3- and 4-input AND gates introduced by `ABY2`. Maybe i'm wrong, i will learn your code carefully, thanks again! :)

Hi @BeStrongok, sorry for the late reply.

You are right, the code is the simplest 2-input adder circuit. The 3- and 4- input circuit is NOT implemented in SPU for now.

IMHO, ABY 2.0 does not bring enough benefits (compared to Cheetah) under the premise of greatly increasing the complexity of the system, so in the case of limited time, we prefer to optimizing Cheetah instead of implement ABY 2.0.

However, the SPU itself is designed as an extensible system and we do not rule out the possibility of introducing ABY2.0. Of course, you are always welcome to contribute new protocols :P

**Comment by BeStrongok** [comment]
> > Thanks for pointing these codes, it's really helpful! But it seems not the 3- and 4-input AND gates introduced by `ABY2`. Maybe i'm wrong, i will learn your code carefully, thanks again! :)
> 
> Hi @BeStrongok, sorry for the late reply.
> 
> You are right, the code is the simplest 2-input adder circuit. The 3- and 4- input circuit is NOT implemented in SPU for now.
> 
> IMHO, ABY 2.0 does not bring enough benefits (compared to Cheetah) under the premise of greatly increasing the complexity of the system, so in the case of limited time, we prefer to optimizing Cheetah instead of implement ABY 2.0.
> 
> However, the SPU itself is designed as an extensible system and we do not rule out the possibility of introducing ABY2.0. Of course, you are always welcome to contribute new protocols :P

Yes, i think these generic primitives may be utilized to optimize the existed protocols. 
I will try to implement the 3- and 4- input circuit and to see if it can bring any benefits, the paper and wonderful implementation of `Cheetah ` is the place where I will focus on learning. 
Thanks for the helpful reply! :)
